# **Descriptive Analytics**

In [4]:
import pandas as pd
import numpy as np

df = pd.read_csv('NYC_Taxi_Cleaned_Analysis_Ready.csv')

In [5]:
display(df.describe())

,VendorID,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,duration_min,speed_mph,pickup_hour
count,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.189311e+06,3.946698e+06,3.946698e+06,3.946698e+06,3.946698e+06
mean,1.787345e+00,1.233373e+00,3.441244e+00,2.710091e+00,1.620718e+02,1.611536e+02,9.466660e-01,1.936859e+01,1.307855e+00,4.963577e-01,3.028975e+00,5.057680e-01,9.830288e-01,2.813628e+01,1.861057e+00,1.493874e-01,5.504911e-01,1.697423e+01,1.102928e+01,1.420994e+01
std,4.194354e-01,6.595587e-01,4.121952e+00,1.264134e+01,6.562439e+01,7.000420e+01,6.359776e-01,1.576509e+01,1.842253e+00,4.729320e-02,3.787247e+00,1.986163e+00,1.288213e-01,1.988695e+01,1.090462e+00,4.889902e-01,3.314068e-01,1.339776e+01,6.144650e+00,5.989917e+00
min,1.000000e+00,0.000000e+00,1.100000e-01,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,1.000000e-02,-1.739000e+01,-7.810000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.010000e+00,0.000000e+00,0.000000e+00,-7.500000e-01,1.000000e+00,1.000000e+00,0.000000e+00
25%,2.000000e+00,1.000000e+00,1.100000e+00,1.000000e+00,1.220000e+02,1.070000e+02,1.000000e+00,9.300000e+00,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,1.000000e+00,1.656000e+01,0.000000e+00,0.000000e+00,0.000000e+00,8.266667e+00,7.082405e+00,1.000000e+01
50%,2.000000e+00,1.000000e+00,1.900000e+00,1.000000e+00,1.610000e+02,1.620000e+02,1.000000e+00,1.420000e+01,0.000000e+00,5.000000e-01,2.450000e+00,0.000000e+00,1.000000e+00,2.186000e+01,2.500000e+00,0.000000e+00,7.500000e-01,1.338333e+01,9.486166e+00,1.500000e+01
75%,2.000000e+00,1.000000e+00,3.700000e+00,1.000000e+00,2.330000e+02,2.330000e+02,1.000000e+00,2.260000e+01,2.500000e+00,5.000000e-01,4.070000e+00,0.000000e+00,1.000000e+00,3.085000e+01,2.500000e+00,0.000000e+00,7.500000e-01,2.116667e+01,1.303040e+01,1.900000e+01
max,6.000000e+00,9.000000e+00,9.906000e+01,9.900000e+01,2.630000e+02,2.630000e+02,4.000000e+00,1.334020e+03,1.500000e+01,2.214000e+01,5.500000e+02,1.158700e+02,1.000000e+00,1.345710e+03,2.500000e+00,1.750000e+00,1.500000e+00,1.799667e+02,8.000000e+01,2.300000e+01


In [6]:
display(df.head(10))

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,duration_min,speed_mph,pickup_hour,day_of_week,is_weekend
0,2,2025-01-18 20:53:30,2025-01-18 21:00:47,1,0.97,1,N,238,166,1,...,1.0,13.32,0.0,0.0,0.00,7.283333,7.990847,20,Saturday,True
1,1,2025-01-25 11:12:51,2025-01-25 11:17:57,1,0.60,1,N,50,48,2,...,1.0,10.55,2.5,0.0,0.75,5.100000,7.058824,11,Saturday,True
2,1,2025-01-21 15:09:31,2025-01-21 15:19:02,1,0.80,1,N,236,237,1,...,1.0,16.65,2.5,0.0,0.00,9.516667,5.043783,15,Tuesday,False
3,2,2025-01-11 22:25:45,2025-01-11 22:34:22,2,1.93,1,N,231,68,1,...,1.0,20.37,2.5,0.0,0.75,8.616667,13.439072,22,Saturday,True
4,2,2025-01-04 23:37:07,2025-01-04 23:45:58,1,4.44,1,N,137,88,0,...,1.0,35.12,0.0,NaN,0.00,8.850000,30.101695,23,Saturday,True
5,1,2025-01-29 00:31:22,2025-01-29 00:38:27,1,1.10,1,N,140,263,1,...,1.0,15.48,2.5,0.0,0.00,7.083333,9.317647,0,Wednesday,False
6,2,2025-01-30 19:19:28,2025-01-30 19:30:09,1,1.33,1,N,170,230,2,...,1.0,17.95,2.5,0.0,0.75,10.683333,7.469579,19,Thursday,False
7,2,2025-01-23 15:59:29,2025-01-23 16:10:38,1,1.25,1,N,43,43,1,...,1.0,16.15,2.5,0.0,0.75,11.150000,6.726457,15,Thursday,False
8,1,2025-01-23 11:27:08,2025-01-23 11:43:02,1,1.70,1,N,100,113,1,...,1.0,23.55,2.5,0.0,0.75,15.900000,6.415094,11,Thursday,False
9,1,2025-01-19 15:42:39,2025-01-19 15:53:35,1,1.20,1,N,163,141,1,...,1.0,16.15,2.5,0.0,0.75,10.933333,6.585366,15,Sunday,True


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3946698 entries, 0 to 3946697
Data columns (total 25 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   VendorID               int64  
 1   tpep_pickup_datetime   object 
 2   tpep_dropoff_datetime  object 
 3   passenger_count        int64  
 4   trip_distance          float64
 5   RatecodeID             int64  
 6   store_and_fwd_flag     object 
 7   PULocationID           int64  
 8   DOLocationID           int64  
 9   payment_type           int64  
 10  fare_amount            float64
 11  extra                  float64
 12  mta_tax                float64
 13  tip_amount             float64
 14  tolls_amount           float64
 15  improvement_surcharge  float64
 16  total_amount           float64
 17  congestion_surcharge   float64
 18  Airport_fee            float64
 19  cbd_congestion_fee     float64
 20  duration_min           float64
 21  speed_mph              float64
 22  pickup_hour       

In [8]:
print(df.isnull().sum())

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count               0
trip_distance                 0
RatecodeID                    0
store_and_fwd_flag            0
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge          0
Airport_fee              757387
cbd_congestion_fee            0
duration_min                  0
speed_mph                     0
pickup_hour                   0
day_of_week                   0
is_weekend                    0
dtype: int64


In [9]:
print(df.isnull().mean() * 100)

VendorID                  0.000000
tpep_pickup_datetime      0.000000
tpep_dropoff_datetime     0.000000
passenger_count           0.000000
trip_distance             0.000000
RatecodeID                0.000000
store_and_fwd_flag        0.000000
PULocationID              0.000000
DOLocationID              0.000000
payment_type              0.000000
fare_amount               0.000000
extra                     0.000000
mta_tax                   0.000000
tip_amount                0.000000
tolls_amount              0.000000
improvement_surcharge     0.000000
total_amount              0.000000
congestion_surcharge      0.000000
Airport_fee              19.190397
cbd_congestion_fee        0.000000
duration_min              0.000000
speed_mph                 0.000000
pickup_hour               0.000000
day_of_week               0.000000
is_weekend                0.000000
dtype: float64


In [10]:
duplicate_count = df.duplicated().sum()
print(f"Total duplicate rows: {duplicate_count}")

Total duplicate rows: 0


In [11]:
display(df.min(numeric_only=True))

,0
VendorID,1
passenger_count,0
trip_distance,0.11
RatecodeID,1
PULocationID,1
DOLocationID,1
payment_type,0
fare_amount,0.01
extra,-17.39
mta_tax,-7.81


In [12]:
display(df.max(numeric_only=True))

,0
VendorID,6
passenger_count,9
trip_distance,99.06
RatecodeID,99
PULocationID,263
DOLocationID,263
payment_type,4
fare_amount,1334.02
extra,15.0
mta_tax,22.14


In [13]:
percentiles = [0.10, 0.25, 0.50, 0.75, 0.90, 0.99]

df_numeric = df.select_dtypes(include=['int', 'float'])

display(df_numeric.quantile(percentiles))

,VendorID,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,duration_min,speed_mph,pickup_hour
0.10,1.0,1.0,0.70,1.0,68.0,54.0,0.0,7.2,0.00,0.5,0.00,0.00,1.0,13.44,0.0,0.00,0.00,5.266667,5.296863,6.0
0.25,2.0,1.0,1.10,1.0,122.0,107.0,1.0,9.3,0.00,0.5,0.00,0.00,1.0,16.56,0.0,0.00,0.00,8.266667,7.082405,10.0
0.50,2.0,1.0,1.90,1.0,161.0,162.0,1.0,14.2,0.00,0.5,2.45,0.00,1.0,21.86,2.5,0.00,0.75,13.383333,9.486166,15.0
0.75,2.0,1.0,3.70,1.0,233.0,233.0,1.0,22.6,2.50,0.5,4.07,0.00,1.0,30.85,2.5,0.00,0.75,21.166667,13.030404,19.0
0.90,2.0,2.0,8.70,1.0,239.0,239.0,2.0,38.7,4.25,0.5,6.65,0.00,1.0,50.51,2.5,0.00,0.75,32.300000,18.941860,21.0
0.99,2.0,4.0,19.27,99.0,263.0,263.0,4.0,70.0,7.50,0.5,17.19,6.94,1.0,101.65,2.5,1.75,0.75,69.383333,33.356289,23.0


In [14]:
display(df.nunique())

,0
VendorID,3
tpep_pickup_datetime,3614373
tpep_dropoff_datetime,3612982
passenger_count,10
trip_distance,3521
RatecodeID,7
store_and_fwd_flag,2
PULocationID,258
DOLocationID,257
payment_type,5


In [15]:
print(f"Total Rows: {df.shape[0]}")
print(f"Total Columns: {df.shape[1]}")

Total Rows: 3946698
Total Columns: 25


## Initial Observations & Notable Findings

### 1. Dataset Scale & Volume
* **High Transaction Volume:** The dataset contains approximately **3.94 million records** (3,946,698 trips), providing a dense sample size across all operating hours and zones.
* **Granularity:** Records capture individual trip events containing pickup/drop-off spatial identifiers, metered timestamps, fee breakdowns, and payment attributes.

### 2. Missing Value Patterns & Structural Gaps
* **Selective Null Clustering:** Over **757,000 records** exhibit missing values in `Airport_fee`. This pattern corresponds structurally to standard street hails that do not originate at airport terminals (JFK or LaGuardia).
* **Unrecorded Operational Expenses:** The dataset does not track vehicle operating expenditures (fuel consumption, vehicle depreciation, maintenance, or vehicle lease fees).
* **Payment-Restricted Tipping:** Recorded values in `tip_amount` are populated exclusively for credit card transactions (`payment_type == 1`); cash tips are systematically unrecorded and register as zero.

### 3. Data Representation & Formatting Nuances
* **Nominal IDs as Numeric Types:** Identifier columns (such as `VendorID`, `RatecodeID`, and `PULocationID`) are stored as numeric/floating-point types rather than categorical variables.
* **Distorted Descriptive Statistics:** Standard summary methods calculate mathematically invalid metrics for identifiers (e.g., a mean `VendorID` of 1.79) and render counts in scientific notation (`3.946698e+06`).
* **Variable Interaction Sensitivity:** Mathematical comparisons and column-wise evaluations are sensitive to mixed column data types across boolean flags and numeric values.

### 4. Revenue Composition, Anomalies & Fee Friction
* **Negative Charge Artifacts:** Descriptive statistics (`.min()` and `.describe()`) revealed negative values in certain fee columns, specifically `extra` (-17.39), `mta_tax` (-7.81), and `cbd_congestion_fee` (-0.75). These anomalies logically represent meter dispute corrections, refunded fares, or system resets rather than standard operational trips.
* **Gross Revenue Inflation:** Raw fare totals (`total_amount`) reflect gross passenger charges rather than driver take-home earnings due to stacked regulatory surcharges (`cbd_congestion_fee`, `congestion_surcharge`, `mta_tax`, `improvement_surcharge`, and `Airport_fee`).
* **Fee Disproportionality:** Flat-rate fees represent a significantly higher proportion of short, low-fare trips compared to long-distance fares.
* **Congestion and Speed Variance:** Wide spreads in `duration_min` and `speed_mph` indicate substantial traffic friction across different pickup locations, suggesting that trip distance alone does not reliably predict trip duration or occupied earning efficiency.

# **Analytical Base Table (ABT)**

In [16]:
# 1. Clean missing fees
df['Airport_fee'] = df['Airport_fee'].fillna(0)

# 2. Revenue & Fee Impact Metrics (For Q1, Q2, Q3)
# Calculate total fees and the true net revenue
df['total_fees'] = df['cbd_congestion_fee'] + df['congestion_surcharge'] + df['Airport_fee']
df['net_revenue'] = df['total_amount'] - df['total_fees']

# Calculate Net Revenue Per Occupied Driving Hour
df['net_revenue_per_hour'] = (df['net_revenue'] / df['duration_min']) * 60

# Calculate the % of gross revenue eaten by fees (Extent of reduction for Q2)
df['fee_impact_pct'] = (df['total_fees'] / df['total_amount']) * 100

# 3. Engineer Fare Tiers & Trip Length Thresholds (For Q2)
# Using qcut to divide fares and distances into 3 equal-sized distribution buckets
df['fare_tier'] = pd.qcut(df['total_amount'], q=3, labels=['Low Fare', 'Medium Fare', 'High Fare'])
df['trip_length_tier'] = pd.qcut(df['trip_distance'], q=3, labels=['Short Trip', 'Medium Trip', 'Long Trip'])

# 4. Engineer Traffic Congestion Levels (For Q3)
# Defining traffic by speed limits/averages (0-8mph = Heavy, 8-15mph = Moderate, 15mph+ = Light)
speed_bins = [0, 8, 15, 100]
speed_labels = ['Heavy Traffic', 'Moderate Traffic', 'Light Traffic']
df['congestion_level'] = pd.cut(df['speed_mph'], bins=speed_bins, labels=speed_labels)

# 5. Preview the variables needed for final analysis
cols_to_show = [
    'day_of_week', 'pickup_hour', 'PULocationID',
    'fare_tier', 'trip_length_tier', 'congestion_level',
    'fee_impact_pct', 'net_revenue_per_hour'
]

print("--- ABT ---")
display(df[cols_to_show].head(10))

--- ABT ---


,day_of_week,pickup_hour,PULocationID,fare_tier,trip_length_tier,congestion_level,fee_impact_pct,net_revenue_per_hour
0,Saturday,20,238,Low Fare,Short Trip,Heavy Traffic,0.000000,109.729977
1,Saturday,11,50,Low Fare,Short Trip,Heavy Traffic,30.805687,85.882353
2,Tuesday,15,236,Low Fare,Short Trip,Heavy Traffic,15.015015,89.211909
3,Saturday,22,231,Medium Fare,Medium Trip,Moderate Traffic,15.954836,119.210832
4,Saturday,23,137,High Fare,Long Trip,Light Traffic,0.000000,238.101695
5,Wednesday,0,140,Low Fare,Short Trip,Moderate Traffic,16.149871,109.948235
6,Thursday,19,170,Low Fare,Short Trip,Heavy Traffic,18.105850,82.558502
7,Thursday,15,43,Low Fare,Short Trip,Heavy Traffic,20.123839,69.417040
8,Thursday,11,100,Medium Fare,Medium Trip,Heavy Traffic,13.800425,76.603774
9,Sunday,15,163,Low Fare,Short Trip,Heavy Traffic,20.123839,70.792683


In [17]:
df.to_csv('NYC_Taxi_ABT.csv', index=False)